pkg/serviceaccounttoken/secret.go

```diff
@@ -117,13 +127,22 @@ func serviceAccountSecretPrefix(sa *v1.ServiceAccount) string {
 	return fmt.Sprintf("%s-token-", sa.Name)
 }
 
-// ServiceAccountSecretName returns the secret name for the given Service Account.
+// ServiceAccountSecret returns the secret for the given Service Account.
 // If there are more than one, it returns the first.
-func ServiceAccountSecretName(sa *v1.ServiceAccount) string {
-	if len(sa.Secrets) < 1 {
-		return ""
+func ServiceAccountSecret(sa *v1.ServiceAccount, secretLister secretLister) (*v1.Secret, error) {
+	if sa == nil {
+		return nil, fmt.Errorf("cannot get secret for nil service account")
+	}
+	secrets, err := secretLister(sa.Namespace, labels.SelectorFromSet(map[string]string{
```

Comment:
Now that we are using the lister to perform this operation, I think that it would be a good idea to implement (at least some) of the following items:
- We should loop over all of the existing secrets, and check if `isSecretForServiceAccount` here. If we find one that matches, we return that one
- If we don't find one, we return nil
  - This will allow us to avoid checking if the returned secret is for a ServiceAccount above
- Optionally, we can delete all secrets which match this label, but aren't for the serviceAccount
  - This helps to avoid issues with users adding secrets that claim they are for the service account, but really aren't.

I think this would also help cases (impersonation) where it calls this function directly rather than calling `Ensure` - in these cases we could return a token that this code seems to know is invalid, but it doesn't convey this fact in the docstring or provide a way to recover if that happens (since it will always get the first invalid value).

Reply:
:+1: